### PR TITLE
Add MultiLockHandler and MultiSemaphore helpers

### PR DESCRIPTION
**Pull request by kufd**
I have added 2 helpers MultiLockHandler and MultiSemaphore which i use in my projects.
So i think it will be usefull for other developers

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
-/.composer.lock
+/composer.lock
 /build/
 /consul-configuration/data-dir/
 /vendor/
+/.idea
```
Comment (fabpot) [inline .gitignore]: This one should be removed. This should be in your personal global ignore list instead.

File: Consul/Helper/MultiLockHandler.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace SensioLabs\Consul\Helper;
+
+use SensioLabs\Consul\Services\KV;
+use SensioLabs\Consul\Services\Session;
+
+/**
+ * Class MultiLockHandler
+ * @package ETWater\ESP\CrawlerBundle\Service
```
Comment (fabpot) [inline Consul/Helper/MultiLockHandler.php]: The `@package` annotation are not needed and must be removed.

File: Consul/Helper/MultiLockHandler.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace SensioLabs\Consul\Helper;
+
+use SensioLabs\Consul\Services\KV;
+use SensioLabs\Consul\Services\Session;
+
+/**
+ * Class MultiLockHandler
```
Comment (fabpot) [inline Consul/Helper/MultiLockHandler.php]: This comment does not add any value, so it can be removed.

**Comment by kufd** [comment]
I have fixed all issues from comments

File: Consul/Helper/MultiLockHandler.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * @return bool
+     */
+    public function lock()
+    {
+        $result = true;
+
+        // Start a session
+        $session = $this->session->create('{"LockDelay":0, "TTL": "'.$this->ttl.'s"}')->json();
```
Comment (lyrixx) [inline Consul/Helper/MultiLockHandler.php]: May be it's better to use `json_encode` here. For example, `ttl` value is never controlled...

File: Consul/Helper/MultiLockHandler.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        $this->sessionId = $session['ID'];
+
+        $lockedResources = [];
+
+        foreach ($this->resources as $resource) {
+            // Lock a key / value with the current session
+            $lockAcquired = $this->kv->put($this->lockPath.$resource, (string) '', ['acquire' => $this->sessionId])->json();
```
Comment (lyrixx) [inline Consul/Helper/MultiLockHandler.php]: so need to cast `''` as string, it's already a string.

File: Consul/Helper/MultiLockHandler.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        if (!$result) {
+            $this->releaseResources($lockedResources);
+        }
+
+        return $result;
+    }
+
+    /**
+     *
+     */
```
Comment (lyrixx) [inline Consul/Helper/MultiLockHandler.php]: should be remove.

File: Consul/Helper/MultiLockHandlerFactory.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        $this->kv = $kv;
+        $this->lockPath = $lockPath;
+    }
+
+    /**
+     * @param array $resources
+     * @param int   $ttl
+     * @return MultiLockHandler
+     */
+    public function getMultiLockHandler(array $resources, $ttl = 60)
```
Comment (lyrixx) [inline Consul/Helper/MultiLockHandlerFactory.php]: I would call this method `createMultiLockHandler`

File: Consul/Helper/MultiLockHandler.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace SensioLabs\Consul\Helper;
+
+use SensioLabs\Consul\Services\KV;
+use SensioLabs\Consul\Services\Session;
+
+class MultiLockHandler
+{
+    /**
+     * @var array
```
Comment (lyrixx) [inline Consul/Helper/MultiLockHandler.php]: Can you put all anotation like that on one like ? `/** @var array */`

File: Consul/Helper/MultiSemaphore.php [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                $result = false;
+            } else {
+                $lockedResources[] = $resource;
+
+                $semaphoreMetaDataValue = [
+                    'limit' => $resource->limit,
+                    'sessions' => [],
+                ];
+
+                //get actuall metadata
```
Comment (lyrixx) [inline Consul/Helper/MultiSemaphore.php]: missing space after `//`

File: Consul/Helper/MultiSemaphore.php [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    public function acquire()
+    {
+        if ($this->sessionId) {
+            throw new \Exception('Resources are acquired.');
+        }
+
+        $result = true;
+
+        // Start a session
+        $session = $this->session->create('{"Name": "semaphor", "LockDelay":0, "TTL": "'.$this->ttl.'s"}')->json();
```
Comment (lyrixx) [inline Consul/Helper/MultiSemaphore.php]: `json_encode` ?

File: Consul/Helper/MultiSemaphore.php [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        $this->sessionId = $session['ID'];
+
+        $lockedResources = [];
+        foreach ($this->resources as $resource) {
+            if (false === $this->kv->put($this->getResourceKey($resource, $this->sessionId), '', ['acquire' => $this->sessionId])->json()) {
+                $result = false;
+            } else {
```
Comment (lyrixx) [inline Consul/Helper/MultiSemaphore.php]: I order to make the code less neested, I would use continue  in the if part, then remove the else and unindent everything.

File: Consul/Helper/MultiSemaphoreFactory.php [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        $this->kv = $kv;
+        $this->keyPrefix = $keyPrefix;
+    }
+
+    /**
+     * @param Resource[] $resources
+     * @param int        $ttl
+     * @return MultiSemaphoreInterface
+     */
+    public function getMultiSemaphore(array $resources, $ttl = 60)
```
Comment (lyrixx) [inline Consul/Helper/MultiSemaphoreFactory.php]: `createMultiSemaphore`

File: Consul/Helper/MultiSemaphoreFactory.php [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $result = null;
+
+        if (empty($resources)) {
```
Comment (lyrixx) [inline Consul/Helper/MultiSemaphoreFactory.php]: `if(!$resources)`

**Comment by lyrixx** [comment]
Good work !

File: Consul/Helper/MultiLockHandler.php (hunk 5 above)
Comment (kufd) [inline Consul/Helper/MultiLockHandler.php]: Fixed

File: Consul/Helper/MultiLockHandler.php (hunk 6 above)
Comment (kufd) [inline Consul/Helper/MultiLockHandler.php]: Removed

File: Consul/Helper/MultiLockHandlerFactory.php (hunk 7 above)
Comment (kufd) [inline Consul/Helper/MultiLockHandlerFactory.php]: Renamed

File: Consul/Helper/MultiLockHandler.php (hunk 8 above)
Comment (kufd) [inline Consul/Helper/MultiLockHandler.php]: Done

File: Consul/Helper/MultiSemaphore.php (hunk 9 above)
Comment (kufd) [inline Consul/Helper/MultiSemaphore.php]: fixed all occurencies

File: Consul/Helper/MultiSemaphoreFactory.php (hunk 12 above)
Comment (kufd) [inline Consul/Helper/MultiSemaphoreFactory.php]: done

File: Consul/Helper/MultiSemaphoreFactory.php (hunk 13 above)
Comment (kufd) [inline Consul/Helper/MultiSemaphoreFactory.php]: done

File: Consul/Helper/MultiSemaphore.php (hunk 11 above)
Comment (kufd) [inline Consul/Helper/MultiSemaphore.php]: I can't do it because below if/else if check variable $resuld and do break if it is false

File: Consul/Helper/MultiLockHandler.php (hunk 4 above)
Comment (kufd) [inline Consul/Helper/MultiLockHandler.php]: json_encode will not protect us from wrong ttl value
So i have added validation for ttl in constructor
Do you think it is fine?

File: Consul/Helper/MultiSemaphore.php (hunk 10 above)
Comment (kufd) [inline Consul/Helper/MultiSemaphore.php]: I have done the same changes as for semaphore

**Comment by kufd** [comment]
I have made changes for all issues from comments

File: Consul/Helper/MultiSemaphore.php (hunk 11 above)
Comment (lyrixx) [inline Consul/Helper/MultiSemaphore.php]: so you can break right now ?

Comment (kufd) [inline Consul/Helper/MultiSemaphore.php]: There is another case in else block when i set false to result
And for me it is better to have only one break command inside foreach instead of heaving 2 break commands.

**Comment by lyrixx** [comment]
@kufd Do you still need this PR? (sorry for the HUGE delay :) )

**Comment by kufd** [comment]
@lyrixx I do not need it

**Comment by lyrixx** [comment]
Can you 

* rebase
* remove all factories
* remove all interfaces

Thanks

**Comment by kufd** [comment]
@lyrixx would you like to merge this request?
I thought you will just close it

However, if you would like to merge the pull request I will prepare it.

**Comment by kufd** [comment]
I will prepare this PR i a few days

**Comment by kufd** [comment]
@lyrixx, i updated the pull request
could you, please, review it

File: src/ConsulResponse.php [hunk 14]
```diff
@@ -30,8 +34,13 @@ public function getStatusCode(): int
         return $this->status;
     }
 
-    public function json(): ?array
```
Comment (lyrixx) [inline src/ConsulResponse.php:33]: Why did you remove the typehint?

File: src/ConsulResponse.php [hunk 15]
```diff
@@ -30,8 +34,13 @@ public function getStatusCode(): int
...
+    public function json()
     {
         return json_decode($this->body, true, 512, \JSON_THROW_ON_ERROR);
     }
+
+    public function isSuccessful(): bool
+    {
+        return \in_array($this->status, [self::HTTP_OK, self::HTTP_CREATED, self::HTTP_NO_CONTENT], true);
```
Comment (lyrixx) [inline src/ConsulResponse.php:44]: I'm not sure theses constant are useful.

You can replace this with 
```suggestion
        return $this->status >= 200 && $this->status < 300;
```

File: src/Helper/MultiLockHandler.php [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+class MultiLockHandler
+{
+    private array $resources;
+    private int $ttl;
+    private Session $session;
+    private KV $kv;
+    private string $sessionId;
+    private string $lockPath;
+
+    public function __construct(array $resources, int $ttl, Session $session, KV $kv, $lockPath)
```
Comment (lyrixx) [inline src/Helper/MultiLockHandler.php:17]:
```suggestion
    public function __construct(array $resources, int $ttl, Session $session, KV $kv, string $lockPath)
```

File: src/Helper/MultiLockHandler.php [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+                break;
+            }
+
+            $lockedResources[] = $resource;
+        }
+
+        if (!$result) {
+            $this->releaseResources($lockedResources);
+        }
```
Comment (lyrixx) [inline src/Helper/MultiLockHandler.php:50]: I would use a `try { } finally { if (!$result) {$this->releaseResources($lockedResources); }`

to ensure you remove everything when something goes wrong.

File: src/Helper/MultiLockHandler.php [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+    public function renew(): bool
+    {
+        return $this->session->renew($this->sessionId)->isSuccessful();
+    }
+
+    public function getResources(): array
+    {
+        return $this->resources;
+    }
```
Comment (lyrixx) [inline src/Helper/MultiLockHandler.php:68]: Is this really useful?

**Review by lyrixx — COMMENTED** [review]
Can you add an entry in the CHANGELOG.md, and a bit of documentation in the README.md

Thanks

File: src/ConsulResponse.php (hunk 15 above)
Comment (kufd) [reply src/ConsulResponse.php]: done

File: src/Helper/MultiLockHandler.php (hunk 17 above)
Comment (kufd) [reply src/Helper/MultiLockHandler.php]: done

File: src/Helper/MultiLockHandler.php (hunk 16 above)
Comment (kufd) [reply src/Helper/MultiLockHandler.php]: done

File: src/Helper/MultiLockHandler.php (hunk 18 above)
Comment (kufd) [reply src/Helper/MultiLockHandler.php:72]: It is hard to predict all cases how developers can use helpers

I suppose it could be usefull

File: src/ConsulResponse.php (hunk 14 above)
Comment (kufd) [reply src/ConsulResponse.php:33]: Because put key request returns true

https://www.consul.io/api-docs/kv#sample-requests

**Comment by kufd** [comment]
@lyrixx I added some documentation and made fixes

File: src/ConsulResponse.php (hunk 14 above)
Comment (lyrixx) [reply src/ConsulResponse.php:33]: oh, thanks

**Comment by lyrixx** [comment]
Here we go 🎉 !

Thanks for your patience and for the hard work.

**Comment by kufd** [comment]
Finally)))
Thank you!
